src/Spago/Command/Build.purs

```diff
@@ -140,42 +97,190 @@ run opts = do
+      dependencyLibs = map (Tuple.uncurry Config.getPackageLocation) dependencyPkgs
+      psaArgs =
+        { libraryDirs: dependencyLibs
+        , color: logOptions.color
+        , jsonErrors: opts.jsonErrors
+        }
+      psaOptions =
+        { strict: fromMaybe Psa.defaultParseOptions.strict workspace.buildOptions.strict
+        , censorBuildWarnings: fromMaybe Psa.defaultParseOptions.censorBuildWarnings workspace.buildOptions.censorBuildWarnings
+        , showSource: fromMaybe Psa.defaultParseOptions.showSource workspace.buildOptions.showSource
+        , censorCodes: maybe Psa.defaultParseOptions.censorCodes NonEmptySet.toSet workspace.buildOptions.censorCodes
+        , filterCodes: maybe Psa.defaultParseOptions.filterCodes NonEmptySet.toSet workspace.buildOptions.filterCodes
+        , statVerbosity: fromMaybe Psa.defaultParseOptions.statVerbosity workspace.buildOptions.statVerbosity
+        , stashFile: do
+            Alternative.guard (not opts.depsOnly)
+            shouldStashWarnings <- workspace.buildOptions.persistWarnings
+            Alternative.guard shouldStashWarnings
+            case workspace.selected of
+              Just p -> Just $ Paths.mkLocalCachesPersistentWarningsFile $ PackageName.print p.package.name
+              Nothing -> Just Paths.localCachesPersistedWarningsEntireWorkspace
+        }
+
+    Psa.psaCompile globs args psaArgs psaOptions
+
+    case workspace.backend of
+      Nothing -> pure unit
+      Just backend -> do
+        logInfo $ "Compiling with backend \"" <> backend.cmd <> "\""
+        logDebug $ "Running command `" <> backend.cmd <> "`"
+        let
+          moreBackendArgs = case backend.args of
+            Just as | Array.length as > 0 -> as
+            _ -> []
+        Cmd.exec backend.cmd (addOutputArgs moreBackendArgs) Cmd.defaultExecOptions >>= case _ of
+          Left err -> do
+            logDebug $ show err
+            die [ "Failed to build with backend " <> backend.cmd ]
+          Right _r ->
+            logSuccess "Backend build succeeded."
+
+    if buildingMultiplePackages then do
+      logDebug $ "Getting purs graph using globs for package: " <> PackageName.print selected.package.name
+      maybeGraph <- Graph.runGraph globs opts.pursArgs
+      case maybeGraph of
+        Nothing ->
+          logWarn $
+            "Due to JSON decoding failure on 'purs graph' output, \
+            \packages that define modules with the same name might not be detected."
+        Just graph -> do
+          let
+            toModulesDefinedByThisPackage :: Purs.ModuleGraphNode -> Maybe (Array (Tuple PackageName FilePath))
+            toModulesDefinedByThisPackage v =
+              [ Tuple selected.package.name v.path ] <$ (String.stripPrefix (String.Pattern selected.path) v.path)
+            modulesDefinedByThisPackage = Map.mapMaybe toModulesDefinedByThisPackage $ unwrap graph
```

Comment:
I was entirely convinced that `purs graph` would fail with `DuplicateModule` in case of modules with the same name, but no, it just takes the last! Or at least I can't get it to fail, maybe I'm not trying hard enough.

Reply:
If we indeed have to do all this bookkeeping, then it should be moved to the Graph module, so we can have a nicely contained `detectDuplicateModules`, or something like that

Reply:
An alternative to doing the bookkeeping is of course to just build everything all together and let it fail with `DuplicateModule` from `purs`

Reply:
> I was entirely convinced that `purs graph` would fail with `DuplicateModule` in case of modules with the same name, but no, it just takes the last! Or at least I can't get it to fail, maybe I'm not trying hard enough.

I think you're running into https://github.com/purescript/purescript/issues/4342

Reply:
Oh no, what a nightmare. Do you think we should fix that now, or paper over it here?

Reply:
It's always better to fix it up-stream, but the problem is that `purs graph` doesn't work until the latest release. So, we'll still need to paper over it here.